test/sqllogictest/github-17616.slt

```diff
@@ -0,0 +1,122 @@
+mode cockroach
+
+# reset
+statement ok
+DROP TABLE IF EXISTS t1 CASCADE
+
+statement ok
+DROP TABLE IF EXISTS t2 CASCADE
+
+# schema
+statement ok
+CREATE TABLE t1(f1 INT, f2 INT, f3 INT)
+
+statement ok
+CREATE TABLE t2(f2 INT, f3 INT)
+
+# data
+statement ok
+INSERT INTO t1 VALUES (1, 2, 3)
+
+statement ok
+INSERT INTO t2 VALUES (2, 3)
+
+query T multiline
+EXPLAIN DECORRELATED PLAN WITH(arity) FOR
+SELECT
+  sq1.*
+FROM
+  (
+    SELECT * FROM t1
+  ) AS sq0,
+  LATERAL(
+    SELECT
+      ( sq0.f2 * f2 ) as c0,
+      ( SELECT 3 ) as c1,
+      ( sq0.f2 * f2 ) * 2 as c2
+    FROM t2
+  ) AS sq1
+----
+Return // { arity: 3 }
+  Project (#3..=#5) // { arity: 3 }
+    Filter true // { arity: 6 }
+      Project (#0..=#2, #4..=#6) // { arity: 6 }
+        Join on=(#1 = #3) // { arity: 7 }
+          Get l0 // { arity: 3 }
+          Project (#0, #3..=#5) // { arity: 4 }
+            Map ((#3 * 2)) // { arity: 6 }
+              Project (#0..=#2, #7, #8) // { arity: 5 }
+                Map ((#0 * #1), #6) // { arity: 9 }
+                  Join on=(#0 = #3 AND #1 = #4 AND #2 = #5) // { arity: 7 }
+                    Get l1 // { arity: 3 }
+                    Union // { arity: 4 }
+                      Get l4 // { arity: 4 }
+                      CrossJoin // { arity: 4 }
+                        Project (#0..=#2) // { arity: 3 }
+                          Join on=(#0 = #3 AND #1 = #4 AND #2 = #5) // { arity: 6 }
+                            Union // { arity: 3 }
+                              Negate // { arity: 3 }
+                                Distinct group_by=[#0..=#2] // { arity: 3 }
+                                  Get l4 // { arity: 4 }
+                              Distinct group_by=[#0..=#2] // { arity: 3 }
+                                Get l2 // { arity: 3 }
+                            Get l2 // { arity: 3 }
+                        Constant // { arity: 1 }
+                          - (null)
+With
+  cte l4 =
+    Union // { arity: 4 }
+      Get l3 // { arity: 4 }
+      Map (error("more than one record produced in subquery")) // { arity: 4 }
+        Project (#0..=#2) // { arity: 3 }
+          Filter (#3 > 1) // { arity: 4 }
+            Reduce group_by=[#0..=#2] aggregates=[count(*)] // { arity: 4 }
+              Get l3 // { arity: 4 }
+  cte l3 =
+    Project (#0..=#3) // { arity: 4 }
+      Map (3) // { arity: 4 }
+        Get l2 // { arity: 3 }
+  cte l2 =
+    Distinct group_by=[#0..=#2] // { arity: 3 }
+      Get l1 // { arity: 3 }
+  cte l1 =
+    CrossJoin // { arity: 3 }
+      Distinct group_by=[#1] // { arity: 1 }
+        Get l0 // { arity: 3 }
+      Get materialize.public.t2 // { arity: 2 }
+  cte l0 =
+    CrossJoin // { arity: 3 }
+      Constant // { arity: 0 }
+        - ()
+      Get materialize.public.t1 // { arity: 3 }
+
+EOF
+
+
+
+query III rowsort
+SELECT
+  sq1.*
+FROM
+  (
+    SELECT * FROM t1
+  ) AS sq0,
+  LATERAL(
+    SELECT
```

Comment:
Before lowering, the `Map` input is of the form

```text
c[0] | ... | c[m]
```

where `c` are input columns.

After lowering, the `Map` input is of the form

```text
o[0] | ... | o[l] | c[0] | ... | c[m] | s[0] | ... | s[n]
```

where `o` are outer columns and `s` contain the results of lowered subqueries defined in the `Map` scalar expressions. 

There are two conditions that need to be satisfied in the input query so it can manifest the bug:

1. At least one subquery (here: `c1`).
2. A `scalars` expression (here: `c0`) that is referenced by a subsequent `scalars` expression (here: `c2`).

Unless you meet condition (1), all `scalars` can be lowered in a single batch regardless of the presence of self-references in the batch or not. This is done by just adding `l` to column references with `level = 0` and substituting references with `level > 0` with one of the `l` entries from the column map. This is already the behavior of `ColumnMap::get`.

Unless you meed condition (2), the `s` suffix in the lowered input is empty. Consequently, even if you have self-referencing entries in your lowered batch, these will be correctly adjusted for by applying `ColumnMap::get`.

Reply:
The fix here basically ensures that `c0` and `c2` will not be lowered in the same step, so we'll never have a situation where a column that comes after the subqueries suffix `s` will be referenced by another column in the same `while !scalars.is_empty() {` lowering step. The `.project` at the end of this loop ensures that the `s` suffix is projected out for the next iteration.